File: solutions/go/twelve-days/1/twelve_days.go
package twelve

import (
	"fmt"
	"strings"
)

var (
	ordinals = []string{
		"first", "second", "third", "fourth", "fifth", "sixth",
		"seventh", "eighth", "ninth", "tenth", "eleventh", "twelfth",
	}
	subjects = []string{
		"a Partridge", "two Turtle Doves", "three French Hens", "four Calling Birds",
		"five Gold Rings", "six Geese-a-Laying", "seven Swans-a-Swimming",
		"eight Maids-a-Milking", "nine Ladies Dancing", "ten Lords-a-Leaping",
		"eleven Pipers Piping", "twelve Drummers Drumming",
	}
)

func joinSubjects(n int) string {
	joinned := ""
	for i := n; i >= 1; i-- {
		joinned += subjects[i] + ", "
	}
	if joinned != "" {
		joinned += "and "
	}
	return joinned + subjects[0]
}

func Song() string {
	allVerses := make([]string, 12)
	for i := 0; i < 12; i++ {
		allVerses[i] = Verse(i + 1)
	}
	return strings.Join(allVerses, "\n")
}

func Verse(v int) string {
	return fmt.Sprintf(
		"On the %s day of Christmas my true love gave to me: %s in a Pear Tree.",
		ordinals[v-1],
		joinSubjects(v-1))
}

File: solutions/lua/twelve-days/1/twelve-days.lua
local ordinals = {
  "first", "second", "third", "fourth", "fifth", "sixth",
  "seventh", "eighth", "ninth", "tenth", "eleventh", "twelfth",
}
local subjects = {
  "a Partridge", "two Turtle Doves", "three French Hens", "four Calling Birds",
  "five Gold Rings", "six Geese-a-Laying", "seven Swans-a-Swimming",
  "eight Maids-a-Milking", "nine Ladies Dancing", "ten Lords-a-Leaping",
  "eleven Pipers Piping", "twelve Drummers Drumming",
}
local function joinSubjects(n)
  if n == 1 then return subjects[1] end
  if n == 2 then return subjects[2] .. ", and " .. subjects[1] end
  return subjects[n] .. ", " .. joinSubjects(n - 1)
end
return {
  recite = function(start_verse, end_verse)
    local verses = setmetatable({}, { __index = table })
    for v = start_verse, end_verse do
      verses:insert(
        ("On the %s day of Christmas my true love gave to me: %s in a Pear Tree."):format(ordinals[v], joinSubjects(v)))
    end
    return verses
  end
}
